KEP-4444: Routing Preference for Services (#4445)
kubernetes/enhancements: keps/sig-network/4444-service-routing-preference/README.md

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes introducing a new field, `routingPreference`, to the
Kubernetes Service spec. It will supersede the functionality currently provided
by the `service.kubernetes.io/topology-mode` annotation and it’s precursor
`topologyKeys` field (which has been deprecated since Kubernetes 1.21)

## Motivation

To be able to understand the motivations for introducing this new field, it’s
important to understand the precursors that shaped this KEP.

### Topology Keys

* This early mechanism allowed users to specify ordered routing preferences (e.g.,
prioritize local nodes, then zone, then anywhere). 

* **Limitations:** However, it lacked flexibility for implementations to make smart
  decisions on its own, like incorporating feedback-based routing optimizations
  (e.g., avoiding overloaded endpoints). The name `topologyKeys` might be
  perceived as overly restrictive, given the desire to incorporate non-topology

>   factors into routing decisions.

Reviewer: Another key thing to mention here is the arbitrary ordering, lengths, and keys here made it very difficult to integrate with systems that might have a concept of more standard settings like preferring same zone routing. With topologyKeys you could theoretically request that an entirely arbitrary topology key be given preference over zone or region.

### Topology Aware Routing and InternalTrafficPolicy

TopologyAwareRouting together with InternalTrafficPolicy were meant to be the
successors of `topologyKeys` and allow implementations to be more flexible.

* TopologyAwareRouting: 
  * Exposes the annotation service.kubernetes.io/topology-mode. When this
    annotation is set to Auto, an implementation specific heuristic is used to
    route the traffic. 
  * **Goal:** The aim with Auto was to allow implementations to be as smart as
    possible and choose what it considered to be the best routing criteria. The
    field didn’t support any other value by design with the thought that the

>     implementation should be able to make the best possible choice for the user.

Reviewer: Maybe mention that this tried to find a balance between safety and same-zone routing, and unfortunately did not achieve either very well.

  * **Limitation:** While designed for intelligent routing, this mode offers less
    user control and can be less predictable. Some users prioritize
    predictability for performance optimization, even when accepting potential

>     endpoint overload in certain situations.

Reviewer: I'd mention that multiple users have filed issues asking why hints weren't applied or working for them. More recent versions added events to help explain what was happening, but there was still a lot of confusion around the feature.

* InternalTrafficPolicy 
  * The spec.internalTrafficPolicy field with the "Local" setting restricts
    traffic to endpoints on the same node as the originating Pod. 
  * **Goal:** With the deprecation of topologyKeys, one of the driving reasons was
    to satisfy the semantic use cases of strictly routing the traffic to the
    same node.
  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.

Note that while the initial proposal of InternalTrafficPolicy proposed a
PreferLocal policy, it was dropped later on. This meant that now
TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
allow users to express a much desired use case from topologyKeys which is

> "prefer node-local, failover to same zone, then route anywhere"

Reviewer: Although this is all 100% correct, I don't think this KEP really addresses that?

  are not required to support all standard heuristics (e.g., PreferZone,
  ProportionalZoneCPU). Even when standard heuristics are supported, their
  precise behavior and interpretation might vary across implementations.

* **Replacement of Traffic Policies:** The new field is complementary to
  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
  their role in enforcing strict traffic locality.

* **Immediate Support for All Possible Heuristics:** The initial implementation
  focuses on a core set of heuristics. Addition of new heuristics (like
  `PreferLocal` for Node local preference) could be explored in future
  refinements.

## Proposal

Add a new field, `routingPreference`, to the Service specification. This field
will serve as preference or hint for the underlying implementation to consider
while making routing decisions. It does not offer strict routing guarantees.

The field will support the following initial values:

* `Default`: Indicates no specific routing preference. The user delegates the
  routing decision to the implementation, allowing it to apply its best-effort
  strategy.

> * `PreferEqualSpread`: Encourages an equal distribution of traffic across

Reviewer: I know the name of the field is still up in the air, but we may need not need the `Prefer` prefix if the field name includes `Preference` as currently proposed.

Reviewer 2: Also agree to drop "Prefer" from the values if it is in the field name

  endpoints, potentially spanning multiple zones (or regions).

> * `PreferZone`: Encourages routing traffic to endpoints within the same zone as

Reviewer: Since this particular approach is aiming to be generic, I'd aim for something more like `PreferClose`, so that the same value could eventually be expanded to included region or other considerations as necessary.

  any heuristic, including standard heuristics. This means the behavior enabled
  by values such as `Default` or `PreferZone` might evolve over time. Such
  refinements could improve the implementation's ability to honor the original
  intent of the heuristic, even if the specific mechanisms change. For example,
  in the case of PreferZone, an implementation might initially route traffic
  within a zone with equal probability. A future improvement could introduce
  load-aware routing within the zone to further optimize performance while still
  adhering to the core principle of zonal preference. The decision of what
  constitutes an "improvement" remains at the discretion of the implementation.

### User Stories

#### Story 1

* **Requirement:** I don't have strong preferences for how my application
  traffic is routed. I prioritize simplicity and trust my Kubernetes
  implementation to optimize traffic distribution.
* **Solution:** Set `routingPreference=Default` (or leave the field unset)
* **Effect:** The Kubernetes implementation will apply its best-effort routing
  strategy based on its design. This strategy might change over time as the

>   implementation evolves. It may load balance across zones or regions regions.

Reviewer: ```suggestion
  implementation evolves. It may load balance across zones or regions.
```

#### Story 2
* **Requirement:** I want my application to primarily receive traffic from
  endpoints within the same zone for performance or cost reasons. However,
  I want to avoid connection failures if no local endpoints are available.
* **Solution:** Set `routingPreference=PreferZone`
* **Effect:** The Kubernetes implementation will aim to prioritize routing
  traffic to endpoints in the same zone as the client. If no endpoints are
  available within the zone, traffic will be routed to other zones. It's
  possible that traffic patterns could lead to endpoints within the preferred
  zone becoming overloaded. Consider other routing strategies or scaling up
  resources within the zone if this becomes a concern.

#### Story 3
* **Requirement:** I prioritize application availability and want to minimize the
  risk of outages due to localized overload. I'm willing to accept potentially
  higher costs associated with cross-zone traffic distribution.

> * **Solution:** Set `routingPreference=PreferSpread`

Reviewer: Nit: This name doesn't match the name you have above.

* **Effect:** The Kubernetes implementation will try to distribute traffic as
  equally as possible across endpoints, potentially spanning multiple zones or
  regions. This can improve resilience but might lead to increased network
  traffic costs.

#### Story 4

> * **Requirement:** As a developer deploying applications across multiple
>   Kubernetes environments, I want a consistent way to express my routing
>   preferences. 

Reviewer: I think what you're going for here is to describe the maintainer of what I'd call a cluster addon. For example, cert-manager, prometheus or really anything on https://artifacthub.io/. Maybe rephrase this to the following:
```suggestion
* **Requirement:** As a developer of a widely deployed cluster addon, I want to be
  able to provide users an easy way to configure my Helm chart and/or deployment
  configuration to enable same-zone routing behavior in a portable way that works
  across many different environments.
```

  be routed, and I know that my chosen implementation supports the desired
  preference. 
* **Solution:** Set `routingPreference=<domain>/<heuristicName>` (where
  `<domain>` and `<heuristicName>` are provided by your implementation).
* **Effect:** The Kubernetes implementation will apply the specified routing
  heuristic. It's important to note that the precise behavior of
  implementation-specific heuristics might vary.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Standard Heuristic Implementation (kube-proxy dataplane)

> kube-proxy based dataplane (along with EndpointSlice controller, within

Reviewer: ```suggestion
kube-proxy (along with EndpointSlice controller, within
```

kube-controller-manager as the control plane) will support the three standard

> routing preferences (`Default`, `PreferEqualSpread`, `PreferZone`).

Reviewer: Maybe worth starting a thread specifically to talk about the names of the values here. Although each of these make sense, I think I'd prefer names that were even a bit more generic, maybe:

- `Default`
- `Even`
- `Close`

This assumes that the name of the field already includes `Preference`. My main goal here would be to move away from `PreferZone` towards something that just allowed for `Close` routing, something that could include region or other keys in the future.

/cc @thockin @danwinship @aojea

#### `Default` and `PreferEqualSpread`
* Initially, kube-proxy will treat the `Default` preference the same as
  `PreferEqualSpread`
* This leverages existing implementation, requiring no major changes.

#### `PreferZone`

> * This preference will be implemented by the use of Hints within EndpointSlices.

Reviewer: Somewhere in this KEP we probably want to mention how we're going to change the existing kube-proxy logic around handling hints: https://github.com/kubernetes/kubernetes/blob/2363cdcc399cbf428210efb2c51575ddcad2b84a/pkg/proxy/topology.go#L151-L155. 

I tend to think that we should remove any check for an annotation or field on the Service and instead honor hints if _every_ endpoint for that Service has a hint specified.

Author: Oh good catch! 
For some reason I was under the impression that we had already made that change but turns out we haven't. I've added a section for this.

implementation choice means that:
* The control plane (kube-controller-manager in this case) is only concerned
  with the field `routingPreference` and populates EndpointSlice hints based
  on its value 
* The data plane (kube-proxy in this case) is only concerned with the Hints
  populated in the EndpointSlice and the fields `internal/externalTrafficPolicy`
  to make routing decisions.
* Neither the control plane nor the data plane looks at the others field.

NOTE: The fact that EndpointSlice hints are not expected to be implemented by
  all data planes is not of concern here, because `routingPreference` is anyways
  supposed to be implementation dependent. Since we know that the kube-proxy
  data plane will respect the Hints, it’s sufficient for the kube-proxy and
  kube-controller-manager based implementations to say that we do support the
  standard heuristics.

### Status Reporting

To provide clear status updates about the routing preferences to the user, the
EndpointSlice controller (which is acting as the control plane) will update the
Service status with the following conditions (inspired by Gateway API conditions)

> * RoutingPreferenceAccepted

Reviewer: I think the only relevant component that could reliably set status on a Service here is the EndpointSlice controller. Given that, I think these are the states we'd want to be able to represent:

1. The EndpointSlice controller sees a value that is not domain-prefixed on a Service and does not know how to handle it
2. The EndpointSlice controller sees a value that it recognizes but can't populate hints due to some other factor (common with the current proportional approach)

It is likely worth representing the positive polarity versions of those as well:
1. The EndpointSlice controller sees a value that it recognizes and has populated hints or determined that it does not need to for the heuristic (Default and PreferEqualSpread would fall into the latter category as proposed)

I think the first item in both of those lists could be handled by this `RoutingPreferenceAccepted` condition that was true or false. Maybe it should be set to "Unknown" for unrecognized/domain-prefixed values?

I think the second item (error populating hints) should probably only appear in the error case. I think your suggestion of `RoutingPreferenceProgrammed` is reasonable here, but since this is very specific to hints, we could instead use something more specific like `HintsPopulated`.

Author: Thanks for giving this a detailed thought, Rob! I mostly have the same thoughts as you on the matter, with the exception of the topic regarding `RoutingPreferenceProgrammed` v/s `HintsPopulated` as the condition name. I was thinking that "HintsPopulated" would be a _reason_ under the `RoutingPreferenceProgrammed` condition. Here's what I had in mind:

* From the user's perspective, the top priority is whether their routing preference has been "somewhat" sent to the data plane. `RoutingPreferenceProgrammed` reflects this directly, while the use of hints may be thought of as an implementation detail (which we will convey through the condition's reason)

* Another reason might spur the debate of being generic vs _YAGNI_ :). But given the previous advantage, in this case I was leaning towards being generic which allows non-hint based implementations to reuse the same condition.

Reviewer: Yep, that makes sense, I think that's a reasonable approach here.

Reviewer: I think the primary reason we needed status on the previous iteration of topology was that there was so much confusion about when it was or was not applied. I'd recommend scrapping status from this stage of the KEP. We can choose to add it in the future if feedback suggests it would be helpful, but I'm hopeful that this will end up being closer to any other Service field where it's not really necessary.

  * **Description:** 
    * This condition indicates whether they `routingPreference` has generated
      some configuration that is assumed to be ready soon in the underlying data
      plane.
    * The EndpointSlice controller will set this status to `True` if it
      successfully populated the EndpointSlice hints based on the
      `routingPreference`.

Note that the EndpointSlice controller and kube-proxy implementation does not
currently provide a condition denoting acknowledgment from the dataplane. In the
future when this is possible, another condition like the following could be
used:

* RoutingPreferenceReady
  * **Type:** `RoutingPreferenceReady`
  * **Description:** Confirms that the dataplane has received the hints,
    acknowledged them, and configured itself accordingly.

#### Condition usage by other implementations

Other implementations supporting `routingPreference` **should** adopt a domain
prefixing strategy for their condition types. This means prefixing condition
types with a domain string (e.g., `my.domain.io/RoutingPreferenceAccepted`) to
prevent conflicts when multiple control planes (like the default EndpointSlice
controller) are present.

> ### Choice of field name

Reviewer 2: Naming debate thread:

I don't love it.  "routing" feels really low-level.  We know this is about locality in practice, but maybe that is too concrete?  I guess, in theory, someone could use different criteria.

readers: ideas welcome

Author: How do folks feel about `trafficDistribution` with `Prefer` keyword in the field values? (Assuming `trafficDistributionPreference` might be too long)

Reviewer 3: agree that "routing" implies the wrong thing

> with `Prefer` keyword in the field values

moving "prefer" from the field name to the values opens up the possibility of having `Require` (or at least "stronger-than-Prefer") values in the future

Reviewer 2: Final names can be argued in impl, but we will need to come back and update this.

Reviewer 4: My tip: come up with really clear name for the feature gate, one that communicates what the feature is and does. Then see if that helps us name the API fields.

<tbody>
  <tr>
    <td>-</td>
    <td>-</td>
    <td>Auto</td>
    <td>routingPreference=Auto</td>
    <td>routingPreference=Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>-</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy=Local</td>
    <td>routingPreference=Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>Local</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy=Local</td>
    <td>InternalTrafficPolicy=Local</td>
  </tr>
</tbody>
</table>

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

> [ ] I/we understand the owners of the involved components may require updates to

Reviewer 2: Gotta check the box

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

> 

Reviewer 2: This section will be of interest

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

> 

Reviewer 4: Not sure that folks are going to like it, but: I'd like to see an alternative discussed where the appropriate heuristic to use is captured in a new API kind.

Let's call this TargetPodHeuristic for now.

So a Service might look like:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: example-service
  namespace: default
spec:
  heuristics:
    - example-heuristic
  selector:
    app.kubernetes.io/name: example
  ports:
    - protocol: TCP
      port: 24
      targetPort: 24
```

and the TargetPodHeuristic for that:
```yaml
apiVersion: example/v1alpha1
kind: TargetPodHeuristic
metadata:
  name: example-heuristic
  namespace: default
spec:
    preferredTargets:
      - weight: 23
        matchLabelKeys:
         - topology.kubernetes.io/zone   # prefer a target in the same zone
         - topology.kubernetes.io/region # prefer a target in the same region
                                         # maybe zone names are only regionally unique
      - weight: 11
        matchLabelKeys:
         - topology.kubernetes.io/region # prefer a target in the same region
    avoidedTargets:
      - weight: 42
        matchTaints:
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoSchedule"
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoExecute"
```

To be clear: we can rule this out if we want to. That's absolutely fine. What I'm calling for is enough of a write-up about why it wouldn't fit / couldn't work / doesn't scale / etc. Let's not leave readers to guess this themselves.

Author: sftim -- That's a reasonable ask. Thanks. I've tried to add something. Hope that conveys the right ideas.

Reviewer 2: This skews back to `topologyKeys` which gave users a ton of control but made alternative implementations very difficult because it was over-precise.

It's a fine line we are walking between defining a full service mesh API and defining too little to be useful.

    implementation should be able to make the best possible choice for the user.
  * **Limitation:** 
      * While designed for intelligent routing, this mode offers less user
        control and can be less predictable. Users filed issues reporting that
        hints weren't being applied or didn't function as expected.
      * The design attempted to strike a balance between safety and preferential
        zone-based routing, but may have fallen short in achieving both
        effectively. Some users prioritize predictability for performance
        optimization, even when accepting potential endpoint overload in certain
        situations.

* InternalTrafficPolicy 
  * The spec.internalTrafficPolicy field with the "Local" setting restricts
    traffic to endpoints on the same node as the originating Pod. 
  * **Goal:** With the deprecation of topologyKeys, one of the driving reasons was
    to satisfy the semantic use cases of strictly routing the traffic to the
    same node.
  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.

Note that while the initial proposal of InternalTrafficPolicy proposed a

> Local policy, it was dropped later on. This meant that now

Author: Thanks for clarifying that, Rob. Yes that was the idea (seems like I mistakenly did a find replace in a some commit :) ). Will fix this.

  enforce deterministic routing paths. It serves as a mechanism for expressing
  hints and preferences that implementations can consider when making routing
  decisions.

* **Mandatory and Uniform Implementation Support:** Kubernetes implementations
  are not required to support all standard heuristics (e.g., `Zone`, `Spread`).
  Even when standard heuristics are supported, their precise behavior and
  interpretation might vary across implementations.

* **Replacement of Traffic Policies:** The new field is complementary to
  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
  their role in enforcing strict traffic locality.

* **Immediate Support for All Possible Heuristics:** The initial implementation
  focuses on a core set of heuristics. Addition of new heuristics (like
  `Local` for Node local preference) could be explored in future
  refinements.

## Proposal

Add a new field, `routingPreference`, to the Service specification. This field
will serve as preference or hint for the underlying implementation to consider
while making routing decisions. It does not offer strict routing guarantees.

The field will support the following initial values:

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer: I don't have a strong opinion here, but can agree with just letting a `nil` value here represent the default lack of preference.

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer 2: This could be haggled in the imp[lementation, too.  Need not block the KEP

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer 2: Can you mark in UNRESOLVED?

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer 3: > Perhaps "Implementation" or "None" is better?

I mean, "None" feels even more like it should be `nil`. If you have a preference about how to route, you specify `routingPreference: Foo`. If you don't have a preference, you just don't specify anything.

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer: > If you don't have a preference, you just don't specify anything

I think we're all agreeing that we don't want users to have to specify anything in this case, the real question is if there's value in us setting a default value here that would show up on reads + be easier to revert to than nil.

> * `Default`: Indicates no specific routing preference. The user delegates the

Reviewer 2: Let's discuss in impl, but I could be OK with the default being nothing (note: "" is something :)

  routing decision to the implementation, allowing it to apply its best-effort
  strategy.
* `Spread`: Encourages an equal distribution of traffic across

>   endpoints, potentially spanning multiple zones (or regions).

Reviewer 3: That sounds like the behavior that is currently expected from kube-proxy (EDIT: yes, basically confirmed below), and yet, by implication, it is _not_ what `Default` means? It seems to me that in past discussions Tim was pretty insistent that kube-proxy _must_ "encourage an equal distribution of traffic across endpoints" by default.

So maybe this should be the `Default`/`nil` behavior, and you should have `Auto` (?) for the "no specific routing preference" behavior.

Reviewer: I think the idea here is to not completely prevent us from having smarter default behavior in the future, but I agree that the default in most cases should be `Spread`. I'd just like to hold out hope that somehow there can be a sufficiently safe "Close" mode that could be enabled by default in the future.

> * `Zone`: Encourages routing traffic to endpoints within the same zone as

Reviewer 2: I hate rehashing this, but I am still unsure.

I don't think we want to be too precise. Calling this "Zone" implies that "Region" and "Node" don't factor.  I don't think we want very many of these.  If "Zone" means anything significantly different than "use zone locality", the name is wrong or the implementation is wrong.

> Now, If we were to choose the heuristic name of Close right now (which at the moment would mean "route traffic within the same zone") and then later on we want to introduce an equivalent to Local (still without any feedback mechanism available to us), I don't think we would be able to say that, henceforth, Close means "route traffic within the same node" because that is increasing the risk. 

I don't think I agree.  If the new definition of "Close" means significantly more risk, the implementation should expose it with a new name.  But is that on US to define?  Adding a standard mode should have a VERY VERY high bar

> * `Zone`: Encourages routing traffic to endpoints within the same zone as

Reviewer 3: I don't think anyone wants a mode that prefers-but-does-not-require same-zone, but which does not also prefer-but-not-require same-region and same-node.

It seems to me like there are three kinds of users/services:
1. "I don't have any special requirements, please just route my services efficiently, and if you can change the routing rules to make it even more efficient in the next release then yay"
2. "I don't have any special requirements, but I don't want to wait for you to make service routing more efficient (or I think I'm smarter than you), so please give me back my `topologyKeys`."
3. "I have specific requirements, which may involve money, and want to be able to enforce those requirements precisely, without worrying that a future version of kube-proxy will optimize around them"

If we aren't going to humor group 2, then groups 1 and 2 collapse together. They basically want "send to the endpoint that best balances (a) low latency, (b) high throughput, (c) the future availability of low-latency high-throughput connections for all clients". Though they can benefit from vaguely-defined routing modes, which will help them express hints like "prioritize latency" (`Close`) or "prioritize throughput/reliability" (`Spread`) so "dumb" proxy implementations can get slightly better behavior, while "smarter" proxy implementations can balance those vague hints against what their own endpoint monitoring is telling them.

But group 3 primarily wants to avoid being charged for cross-zone traffic. If their cluster is in a situation where zone 1's endpoints are getting overloaded and zone 2's endpoints are idle, they don't want Kubernetes to automatically start routing traffic to zone 2 for them, they want Kubernetes to automatically deploy more endpoints to zone 1 for them. So they want modes that are extremely precise about some things, though they can still be vague about other things. (eg, if they say `RequireSameZone`, the proxy is still allowed to try to optimize endpoint balance _after_ it has enforced that constraint).

If we don't care about group 3, then I don't think there's any reason for the API to mention zones at all. It doesn't make sense to have a mode that means "keep traffic within the same zone if it makes sense, but let it go to other zones if that would be better", because that should just be part of `Default`.

> * `Zone`: Encourages routing traffic to endpoints within the same zone as

Reviewer 2: We think almost everyone should choose Default, right?  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."  

The user-story around helm charts is what keeps coming back to my mind.  If I know that I have set up zonal scaling and so forth, it's OK for me to choose "fast" over "good" (because I handled Goodness myself).

So maybe the question is whether this is really a PREFERENCE that implementations may ignore (I know you said prefer zone, but I know from feedback that this is bad) or strict?  If you really really want strict same-zone, maybe you should have a different Service per zone?  That's crappy for LBs, I guess.

So I'll ask again: does "Zone" mean an implementation must ignore information and make "wrong" choices?  Does "Spread" mean it must cross zones sometimes, just to say it approached a random distribution?

> * `Zone`: Encourages routing traffic to endpoints within the same zone as

Reviewer 3: > We think almost everyone should choose Default, right?

In the future, yes, almost everyone should choose Default (while they zip around Mars with their jetpacks). In the short run though, we know nobody is actually doing endpoint monitoring, and we don't have any specific plans to add it to kube-proxy, so hints seem useful. (But we want to define the hints so that future jetpack kubernetes isn't overly constrained by them.)

>  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."

zone is sometimes about "cheap", not just "fast"

> So maybe the question is whether this is really a PREFERENCE that implementations may ignore

There are people who at least claim that they want strict.

>  ("I know you said prefer zone, but I know from feedback that this is bad")

It was bad at the point you gathered the feedback, but what if their helm chart includes a HorizontalPodAutoscaler that is already in the process of fixing that?

We can assume that the service proxy is more clever than the average user, but we can't assume it's more clever than _every_ user...

---

Another (possibly implausible) "strict" example: "clients in the EU region must be routed to endpoints in the EU region, for GDPR compliance". If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections get _dropped_ than be misrouted. (Which is probably not actually what most people want in the "RequireZone" case...)

> * `Zone`: Encourages routing traffic to endpoints within the same zone as

Reviewer: > If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections fail than be misrouted. 

This is a good point and something that the current generation of hints are sadly unprepared to handle since we simply fall back to cluster-wide routing if no endpoints have matching zone hints. I think historically we've said that "RequireZone" and "RequireRegion" were sufficiently niche to be out of scope for this, but it's worth noting that not just are they out of scope, they'd require different implementation details (although of course it could just be new values added to this API in the future).

The very nature of the proposed field name here indicates that it's a preference and not a policy, similar conceptually to "hint" vs "requirement". I think that we might be able to make a case for something like "RequireZone" belonging as an option in internal and external traffic policy fields in the future instead of this preference field, but maybe that's an unnecessary and unhelpful distinction. In that world you'd have the following logic for each type of traffic:

_Is the policy field set to a non-"cluster" value for this type of traffic?_

Yes: Policy has precedence, policy is likely ignored unless it's somehow meaningful within the subset of endpoints the policy results in
No: Use the routing preference specified by this new field

Does that logic make sense? Is it weird to have external vs internal distinctions for traffic policies but not this routing preference?

For reference, what I've described above is pretty similar to [how we handle this today in kube-proxy](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/pkg/proxy/topology.go#L40) with hints, with the difference being that this leaves open the possibility that both a policy and a preference could both be applied for the same service which is not currently possible.

  * The new field doesn't support a routing preference that is exactly similar
    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.
  * If both field and the annotation are set, the annotation will take
    precedence. (However, this behavior is temporary as the annotation will be
    deprecated and removed in future releases)

  **Mitigation:** Properly document the suggested migration paths with
  limitations.

## Design Details

### Standard Heuristic Implementation (kube-proxy dataplane)

kube-proxy (along with EndpointSlice controller, within kube-controller-manager
as the control plane) will support the three standard routing preferences
(`Default`, `Spread`, `Zone`).

#### `Default` and `Spread`
* Initially, kube-proxy will treat the `Default` preference the same as
  `Spread`
* This leverages existing implementation, requiring no major changes.

#### `Zone`
* This preference will be implemented by the use of Hints within EndpointSlices.
* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
  Similarly, we’ll use the same Hints within the EndpointSlice to implement the

>   `Zone` heuristic – the hints will match the zone of the endpoint itself.

Reviewer 3: So how will `routingPreference: Zone` address the aforementioned Limitations of Topology-Aware Hints?

Reviewer: I think the biggest benefit of this simpler approach is the predictability and simplicity of this approach. It will also work equally well for Services with small numbers of endpoints. If I were to summarize the feedback I've gotten about the current approach, it would largely fall into some combination of the following:

1. Why aren't hints being set?
2. How do I ensure that hints will be reliably set?

This approach should avoid both of those common points of frustration.

Reviewer 3: > It seems like CPU-proportional is a more or less failed experiment.

Oh... I see. When the KEP says "the hints will match the zone of the endpoint itself", it's talking about _making a change_ to kcm, not explaining the current behavior of kcm. That is not clear (and is the source of my confusion below as well).

* While it may seem redundant to populate the hints here since kube-proxy can
  already derive the zone hint from the endpoints zone (as they would be the
  same), we will still use this for implementation simply because of the reason
  that it’s easier to implement and provides a better design. Consider an
  alternative implementation where kube-proxy reads
  `routingPreference=Zone` field and then constructs the route rules
  accordingly. This means some extra logic needs to be baked into the kube-proxy
  which could have just as easily been implemented by an already existing

>   extensibility mechanism (i.e. EndpointSlice hints)

Reviewer 3: I don't understand this. It sounds like you are saying "having two fields that will always have the same value is better design than only having one"

  with the field `routingPreference` and populates EndpointSlice hints based
  on its value 
* The data plane (kube-proxy in this case) is only concerned with the Hints
  populated in the EndpointSlice and the fields `internal/externalTrafficPolicy`
  to make routing decisions.
* Neither the control plane nor the data plane looks at the others field.

NOTE: The fact that EndpointSlice hints are not expected to be implemented by
  all data planes is not of concern here, because `routingPreference` is anyways
  supposed to be implementation dependent. Since we know that the kube-proxy
  data plane will respect the Hints, it’s sufficient for the kube-proxy and
  kube-controller-manager based implementations to say that we do support the
  standard heuristics.

### Changes within kube-proxy

**Present behaviour:** kube-proxy only considers EndpointSlice hints for route
programming if the `service.kubernetes.io/topology-aware-hints` annotation is
set to "Auto"

**New behaviour:** Irrespective of what the annotation
`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
set to (or even if they are not set at all), kube-proxy will always consider

> EndpointSlice hints (assuming this feature-gate is enabled).

Reviewer 3: It will "consider" hints regardless of `routingPreference` value? That needs clarification. I assume it "considers" them in very different ways for `Spread` vs `Zone`.

implementation-specific nature of this field and how it affects the routing of
traffic
* Field names that include the word "topology", like `topologyPreference` and
  `topologyRoutingPreference`, were avoided because:
  * Topology word might be a bit vague
  * The actual heuristic might not specifically be tied to the geographical or
    network topology, but might also be considering factors like latency, active
    connections, cpu usage of the backends, etc.
* Use of the word “hint” for names like `trafficRoutingHint` and
  `customRoutingHint` was avoided so as not to confuse with the hint field
  within EndpointSlices which has previously been discussed in relation to such
  routing heuristics. In a sense, the hint field within EndpointSlices is just
  an implementation detail.
* Use of words like “traffic” and “policy” for names like
  trafficRoutingPreference was not favored so that it’s clearly different from
  the `internalTrafficPolicy` and `externalTrafficPolicy` fields which have a
  more fixed behavior.
* Use of the word "selection" for a name like `endpointSelection` were avoided
  so as not to confuse with the actual process of selecting the complete set of

>   pods backing a service.

Reviewer 3: `endpointSelection` was my suggestion before I got to this comment... I'm not sure if I agree or disagree with that objection...

maybe `endpointDistribution`?

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Alternatives

Reviewer 3: At some point in the last discussion, we discovered [`pod.spec.topologySpreadConstraints`](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/), which are a way of saying things like "I want this Deployment deployed such that each zone/node/whatever has the same number of Pods", and there was [a brief discussion](https://github.com/kubernetes/enhancements/pull/3293#issuecomment-1322698198) about how we could use that for topology-aware service routing.

Reviewer: +1 - definitely worth covering that option in this section. My personal opinion is that that API is both too complex for users and too limiting for implementations.

Author: Thanks for the suggestions. I've added a section on `Reuse Pod Topology Spread Constraints for Traffic Distribution`. Additionally, feels like rather than being an either-or situation, it's more appropriate for Pod Topology Spread Constraints to complement the `routingPreference`. Added a section on `Complementary Use of Pod Topology Spread Constraints and routingPreference` which eventually might also make sense in a user facing doc

> My personal opinion is that that API is both too complex for users and too limiting for implementations.

I share the same thoughts. The section on `Granular Routing Controls` below exactly emphasis those downsides.

  refinements.

## Proposal

Add a new field, `routingPreference`, to the Service specification. This field
will serve as preference or hint for the underlying implementation to consider
while making routing decisions. It does not offer strict routing guarantees.

The field will support the following initial values:

* `Default`: Indicates no specific routing preference. The user delegates the
  routing decision to the implementation, allowing it to apply its best-effort
  strategy.
* `Close`: Indicates a preference for routing traffic to endpoints that
  are topologically proximate to the client. The interpretation of
  "topologically proximate" may vary across implementations and could encompass
  endpoints within the same node, rack, zone, or even region.

Implementations SHOULD support the standard values. While some flexibility in
interpretation is permitted, implementations should aim to align their behavior
with the described intent of these preferences as closely as possible.

Implementations may support additional routing heuristics using values of the
form `<domain>/<heuristicName>`. Heuristics without a domain prefix will be

> reserved for potential future standardization.

Reviewer: I'd personally recommend leaving this as an optional future enhancement to limit the scope of this KEP for now. We can always loosen validation to allow this in the future if there's demand for it.

        optimization, even when accepting potential endpoint overload in certain
        situations.

* InternalTrafficPolicy 
  * The spec.internalTrafficPolicy field with the "Local" setting restricts
    traffic to endpoints on the same node as the originating Pod. 
  * **Goal:** With the deprecation of topologyKeys, one of the driving reasons was
    to satisfy the semantic use cases of strictly routing the traffic to the
    same node.
  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.

Note that while the initial proposal of InternalTrafficPolicy proposed a
`PreferLocal` policy, it was dropped later on. This meant that now
TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
allow users to express a much desired use case from topologyKeys which is
"prefer node-local, failover to same zone, then route anywhere" While this
specific behavior is a non-goal for the initial implementation of the
`routingPreference` field, the design allows for the potential introduction of
such a preference in future refinements.

## Goals

* **Enhanced User Control:** Provide users with a knob to specify a preference that

>   MAY influence the traffic routing of a Kubernetes Service.

Reviewer 4: ```suggestion
  **may** influence the traffic routing of a Kubernetes Service.
```

* **Guidance for Implementations:** Offer standard routing preference values
  that implementations can choose to recognize and support, promoting a degree
  of commonality.

* **Flexibility:** Allow implementations to interpret and support standard
  heuristics in a manner that aligns with their capabilities and design.

* **Extensibility:** Enable Kubernetes implementations to introduce and support
  innovative routing strategies (for example, those based on topology, latency,
  or other custom heuristics) that can evolve alongside users needs and
  infrastructure capabilities.

## Non-Goals

* **Strict Routing Guarantees:** The `routingPreference` field will not
  enforce deterministic routing paths. It serves as a mechanism for expressing
  hints and preferences that implementations can consider when making routing

>   decisions.

Reviewer 4: If the guarantees aren't there, are we as a side effect making it harder for Kubernetes to offer those guarantees one day?

* **Mandatory and Uniform Implementation Support:** Kubernetes implementations
  are not required to support all standard heuristics (e.g., `Zone`, `Spread`).
  Even when standard heuristics are supported, their precise behavior and
  interpretation might vary across implementations.

* **Replacement of Traffic Policies:** The new field is complementary to

>   InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute

Reviewer 4: ```suggestion
  `internalTrafficPolicy` and `externalTrafficPolicy`. It does not aim to substitute
```

  their role in enforcing strict traffic locality.

* **Immediate Support for All Possible Heuristics:** The initial implementation
  focuses on a core set of heuristics. Addition of new heuristics (like
  `Local` for Node local preference) could be explored in future
  refinements.

## Proposal

Add a new field, `routingPreference`, to the Service specification. This field
will serve as preference or hint for the underlying implementation to consider
while making routing decisions. It does not offer strict routing guarantees.

The field will support the following initial values:

* `Default`: Indicates no specific routing preference. The user delegates the
  routing decision to the implementation, allowing it to apply its best-effort
  strategy.
* `Spread`: Encourages an equal distribution of traffic across
  endpoints, potentially spanning multiple zones (or regions).
* `Zone`: Encourages routing traffic to endpoints within the same zone as
  the client. If no endpoints are available within the zone, traffic should be
  routed to other zones.

> Implementations are strongly encouraged to support the standard values. While

Reviewer 2: "Strongly encoraged" != required.

Is it OK for an implementation to not support "Zone" at all?  If so, should it accept the value and ignore it, or should it fail to program endpoints at all?

as possible.

Implementations may support additional routing heuristics using values of the
form `<domain>/<heuristicName>`. Heuristics without a domain prefix will be
reserved for potential future standardization.

NOTE: Implementations reserve the right to refine the behavior associated with
  any heuristic, including standard heuristics. This means the behavior enabled
  by values such as `Default` or `Zone` might evolve over time, and some
  evolutions might interpret the heuristic goals slightly differently. For
  example, in the case of `Zone`, an implementation might initially route
  traffic within the zone without considering endpoint overload, while a future
  refinement could introduce feedback mechanisms to detect overload and route
  traffic outside the zone when necessary, optimizing overall performance. The
  decision of what constitutes an "improvement" remains at the discretion of the
  implementation.

### User Stories

#### Story 1

* **Requirement:** I don't have strong preferences for how my application
  traffic is routed. I prioritize simplicity and trust my Kubernetes
  implementation to optimize traffic distribution.

> * **Solution:** Set `routingPreference=Default` (or leave the field unset)

Reviewer 4: (nit)
```suggestion
* **Solution:** Set `routingPreference: Default` (or leave the field unset)
```

* **Effect:** The Kubernetes implementation will apply its best-effort routing
  strategy based on its design. This strategy might change over time as the
  implementation evolves. It may load balance across zones or regions.

#### Story 2
* **Requirement:** I want my application to primarily receive traffic from
  endpoints within the same zone for performance or cost reasons. However,
  I want to avoid connection failures if no local endpoints are available.

> * **Solution:** Set `routingPreference=Zone`

Reviewer 4: (nit)
```suggestion
* **Solution:** Set `routingPreference: Zone`
```

* **Effect:** The Kubernetes implementation will aim to prioritize routing
  traffic to endpoints in the same zone as the client. If no endpoints are
  available within the zone, traffic will be routed to other zones. It's
  possible that traffic patterns could lead to endpoints within the preferred
  zone becoming overloaded. Consider other routing strategies or scaling up

>   resources within the zone if this becomes a concern.

Reviewer 4: For this user story, it sounds as if the user gets to find out if their preference worked once they see the next bill. I predict unhappy finance teams.

#### Story 3
* **Requirement:** I prioritize application availability and want to minimize the
  risk of outages due to localized overload. I'm willing to accept potentially
  higher costs associated with cross-zone traffic distribution.

> * **Solution:** Set `routingPreference=Spread`

Reviewer 4: (nit)
```suggestion
* **Solution:** Set `routingPreference: Spread`
```

* **Effect:** The Kubernetes implementation will try to distribute traffic as
  equally as possible across endpoints, potentially spanning multiple zones or
  regions. This can improve resilience but might lead to increased network
  traffic costs.

#### Story 4
* **Requirement:** As a developer of a widely deployed cluster-addon, I want to
  be able to provide users an easy way to configure my Helm chart and/or
  deployment configuration to enable same-zone routing behavior in a portable
  way that works across many different environments.
* **Solution:** Using one of the available standard values in the
  `routingPreference` field will allow the customer to use the same Helm Chart
  configurations with greater confidence regardless of the underlying Kubernetes
  implementation. This simplifies their deployment process and reduces the
  complexity of managing cross-cluster applications.

#### Story 5
* **Requirement:** I have some other precise preferences for how traffic should
  be routed, and I know that my chosen implementation supports the desired
  preference. 

> * **Solution:** Set `routingPreference=<domain>/<heuristicName>` (where

Reviewer 4: (nit)
```suggestion
* **Solution:** Set `routingPreference: <domain>/<heuristicName>` (where
```

  `<domain>` and `<heuristicName>` are provided by your implementation).
* **Effect:** The Kubernetes implementation will apply the specified routing
  heuristic. It's important to note that the precise behavior of
  implementation-specific heuristics might vary.

> ### Notes/Constraints/Caveats (Optional)

Reviewer 4: (nit)
```suggestion
### Notes/Constraints/Caveats
```

      independently of other zones.

* **Risk:** Users migrating from the `service.kubernetes.io/topology-mode`
  annotation might encounter differences in exact routing behavior:
  * The new field doesn't support a routing preference that is exactly similar
    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.
  * If both field and the annotation are set, the annotation will take
    precedence. (However, this behavior is temporary as the annotation will be
    deprecated and removed in future releases)

  **Mitigation:** Properly document the suggested migration paths with
  limitations.

## Design Details

### Standard Heuristic Implementation (kube-proxy dataplane)

kube-proxy (along with EndpointSlice controller, within kube-controller-manager
as the control plane) will support the three standard routing preferences
(`Default`, `Spread`, `Zone`).

#### `Default` and `Spread`
* Initially, kube-proxy will treat the `Default` preference the same as
  `Spread`
* This leverages existing implementation, requiring no major changes.

#### `Zone`
* This preference will be implemented by the use of Hints within EndpointSlices.

> * We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`

Reviewer 4: (nit)
```suggestion
* We already use Hints to implement `service.kubernetes.io/topology-mode: Auto`
```

  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
  `Zone` heuristic – the hints will match the zone of the endpoint itself.
* While it may seem redundant to populate the hints here since kube-proxy can
  already derive the zone hint from the endpoints zone (as they would be the
  same), we will still use this for implementation simply because of the reason
  that it’s easier to implement and provides a better design. Consider an
  alternative implementation where kube-proxy reads

>   `routingPreference=Zone` field and then constructs the route rules

Reviewer 4: ```suggestion
  `routingPreference: Zone` and then constructs the route rules
```

NOTE: The fact that EndpointSlice hints are not expected to be implemented by
  all data planes is not of concern here, because `routingPreference` is anyways
  supposed to be implementation dependent. Since we know that the kube-proxy
  data plane will respect the Hints, it’s sufficient for the kube-proxy and
  kube-controller-manager based implementations to say that we do support the
  standard heuristics.

### Changes within kube-proxy

**Present behaviour:** kube-proxy only considers EndpointSlice hints for route
programming if the `service.kubernetes.io/topology-aware-hints` annotation is
set to "Auto"

**New behaviour:** Irrespective of what the annotation
`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
set to (or even if they are not set at all), kube-proxy will always consider
EndpointSlice hints (assuming this feature-gate is enabled).

NOTE: The expectation remains that *all* endpoints within an EndpointSlice must
  have corresponding hints for kube-proxy to utilize them. This avoids scenarios
  with partial hints. The reason for this requirement is the same one highlighted in [KEP-2433 Topology
  Aware

>   Hints](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2433-topology-aware-hints/README.md#kube-proxy), i.e. _"This is to provide safer transitions between enabled and disabled states. Without this fallback, endpoints could easily get overloaded as hints were being added or removed from some EndpointSlices but had not yet propagated to all of them."_

Reviewer 4: Is it OK to have multiple EndpointSlices for one Service, where some of those EndpointSlices have hints for all endpoints and other EndpointSlices have no hints? And, if it is OK, is it OK for that to be a lasting situation?

    * Indicates whether a control plane component recognized and successfully
      parsed the `routingPreference` value. A `False` status typically suggests
      a configuration issue (e.g., an unsupported or malformed preference).
    * The EndpointSlice controller will set this status to `True` if it
      recognizes the `routingPreference` value as one it explicitly supports. 

* RoutingPreferenceProgrammed
  * **Type:**  `RoutingPreferenceProgrammed`
  * **Description:** 
    * This condition indicates whether they `routingPreference` has generated
      some configuration that is assumed to be ready soon in the underlying data
      plane.
    * The EndpointSlice controller will set this status to `True` if it
      successfully populated the EndpointSlice hints based on the
      `routingPreference`.

Note that the EndpointSlice controller and kube-proxy implementation does not
currently provide a condition denoting acknowledgment from the dataplane. In the
future when this is possible, another condition like the following could be
used:

* RoutingPreferenceReady
  * **Type:** `RoutingPreferenceReady`
  * **Description:** Confirms that the dataplane has received the hints,

>     acknowledged them, and configured itself accordingly.

Reviewer 4: It'll be tricky to get right if we make that available - it requires aggregating the underlying status from lots of places.

---

I'd call this `RoutingPreferenceHonored`. Because it's a preference, not honoring it is OK - but people still want to know if it did get honored.

  an implementation detail.
* Use of words like “traffic” and “policy” for names like
  trafficRoutingPreference was not favored so that it’s clearly different from
  the `internalTrafficPolicy` and `externalTrafficPolicy` fields which have a
  more fixed behavior.
* Use of the word "selection" for a name like `endpointSelection` were avoided
  so as not to confuse with the actual process of selecting the complete set of
  pods backing a service.

<<[/UNRESOLVED]>>

### Intersection with internal/externalTrafficPolicy

The intersection of the field with `internalTrafficPolicy` and
`externalTrafficPolicy` fields remains the same as the annotation. The following
table borrowed from [KEP-2086: Service Internal Traffic
Policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy)
captures the precedence

<table>
<thead>
  <tr>
    <th>ExternalTrafficPolicy</th>
    <th>InternalTrafficPolicy</th>
    <th>routingPreference</th>
    <th>External Result</th>
    <th>Internal Result</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>-</td>
    <td>-</td>
    <td>Auto</td>

>     <td>routingPreference=Auto</td>
>     <td>routingPreference=Auto</td>
>   </tr>
>   <tr>
>     <td>Local</td>
>     <td>-</td>
>     <td>Auto</td>
>     <td>ExternalTrafficPolicy=Local</td>
>     <td>routingPreference=Auto</td>
>   </tr>
>   <tr>
>     <td>Local</td>
>     <td>Local</td>
>     <td>Auto</td>
>     <td>ExternalTrafficPolicy=Local</td>
>     <td>InternalTrafficPolicy=Local</td>

Reviewer 4: ```suggestion
    <td>routingPreference: Auto</td>
    <td>routingPreference: Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>-</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy: Local</td>
    <td>routingPreference: Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>Local</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy: Local</td>
    <td>InternalTrafficPolicy: Local</td>
```

This is valid YAML (snippets) and is closer to valid JSON than the version using `=`.

  </tr>
</tbody>
</table>

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

No updates required.

##### Unit tests

The main packages that will see major changes due to this enhancement are:
- `k8s.io/kubernetes/vendor/k8s.io/endpointslice/topologycache`: `2024-01-26` -
  `74.8`
- `k8s.io/kubernetes/vendor/k8s.io/endpointslice`: `2024-01-26` - `80.7`

There will be some code refactoring into newer packages which may prevent exact
comparisons, but the updated packages will aim to provide equivalent coverage.

The following packages will also see minor changes:
- `k8s.io/kubernetes/pkg/controller/endpointslice`: `2024-01-26` - `61.4`
- `k8s.io/kubernetes/pkg/proxy`: `2024-01-26` - `69.5`

##### Integration tests

> * Verify that if both the annotation `service.kubernetes.io/topology-mode=Auto`
>   and field `routingPreference=Zone` are configured, precedence is given
>   to the annotation.

Reviewer 4: ```suggestion
* Verify that if both the annotation `service.kubernetes.io/topology-mode: Auto`
  and field `routingPreference: Zone` are configured, precedence is given
  to the annotation.
```

##### e2e tests

> * Verify that EndpointSlice hints are correctly populated when
>   `routingPreference=Zone`
> * Verify through probes that when `routingPreference=Zone`, requests

Reviewer 4: (nit)
```suggestion
* Verify that EndpointSlice hints are correctly populated when
  `routingPreference` is set to `Zone`
* Verify through probes that for `routingPreference: Zone`, requests
```

  originating from a zone which has service pods get sent to a pod in the same
  zone. For requests originating from zones with no service pods, requests
  should not get blackholed and should rather be forwarded to any service pod

>   from the cluster.

Reviewer 4: Can we usefully end-to-end test a _preference_? It seems to me that the only thing we can actually assert is that traffic does not get dropped into a hole.

### Graduation Criteria

#### Alpha

> - Feature implemented behind a feature flag

Reviewer 4: (nit)
```suggestion
- Feature implemented behind a feature gate
```

- Initial e2e tests completed and enabled

### Upgrade / Downgrade Strategy

No special considerations are required for upgrade / downgrade:

Upon upgrade of the EndpointSlice controller (within kube-controller-manager)
and kube-proxy:
* If a Service had both annotation `service.kubernetes.io/topology-mode` and
  field `routingPreference`, then the annotation will take precedence and the
  field will be ignored. 
* If a Service only had the annotation `service.kubernetes.io/topology-mode`,
  then field `routingPreference`'s value will be assumed to be `Default` and the
  behaviour will be the same as above with the annotation taking precedence.

Upon downgrade of EndpointSlice controller (within kube-controller-manager) and
kube-proxy, the `routingPreference` will simply not be considered in any

> decisions.

Reviewer 4: For downgrade, how would the API server handle the `routingPreference` field that it finds on some persisted objects? Any concerns there?

### Version Skew Strategy

Version skews should naturally get handled as per the following behaviour.

* **kube-apiserver:** [Kubernetes Version Skew
  Policies](https://kubernetes.io/releases/version-skew-policy/#supported-version-skew)

>   require that kube-apiserver is atleast at the version of kube-proxy or

Reviewer 4: ```suggestion
  require that kube-apiserver is at least at the version of kube-proxy or
```

  kube-controller-manager. The only valid version skew would mean that a newer
  kube-apiserver serves the new `routingPreference` field but the older
  kube-proxy and kube-controller-manager would silently ignore this field. (No

>   adverse affect, behaviour equivalent to feature being disabled)

Reviewer 4: (nit)
```suggestion
  adverse affect, behaviour equivalent to feature being disabled).
```

complexity. 

A dedicated `routingPreference` fields gets us:

* **Clear Intent:** The `routingPreference` field provides an explicit way for
  users to signal desired traffic patterns, focusing solely on traffic
  distribution.

* **Implementation Flexibility:** Implementations can intelligently incorporate
  Topology Spread Constraints information (if desired) alongside other factors
  like latency, load, or custom heuristics to optimize routing decisions.

#### Complementary Use of Pod Topology Spread Constraints and routingPreference

Rather than having to choose between the two, Pod Topology Spread Constraints
and `routingPreference` can offer slightly better and resilient traffic
distribution when used in conjunction.

* Users can set `routingPreference` to `Zone` to express the preference for
  keeping traffic within the same zone as the client.
* Then, they can configure Pod Topology Spread Constraints to Ensure balanced
  pod distribution across zones, maximizing the likelihood that the
  `routingPreference` can be satisfied and reduce (although not completely
  eliminate) chances of overload for a single zone.

> 

Reviewer 4: Is there an alternative available where we put the complexity into Gateway to ensure that there is a SomethingWeNeedToName (ie, not Service) per zone, and it does, and we're happy?

In this story:
- maybe a SomethingWeNeedToName controller manages the EndpointSlices, rather than kube-controller-manager
- maybe you can still refer to the Service if you want to, for example using [environment variables](https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables) - and the Cluster IP for that Service sends traffic somewhere reasonable.
- maybe we annotate Services when we want SomethingWeNeedToNames set up per zone.

That's not fleshed out; the key ask is: _did we already consider putting the complexity [mostly] out of tree into Gateway, with kube-proxy learning how to integrate and other in-tree code remaining unaware?_

> ## Infrastructure Needed (Optional)
> 
> N/A

Reviewer 4: ```suggestion
```

### Notes/Constraints/Caveats (Optional)

> None.

Reviewer 2: Caveat: This being our third attempt at an API for this, there's a non-zero chance that we STILL didn't get it right, and will have to revisit again.

Caveat: Converting from the annotation to the field is not quite a mechanical change, which means users will have to think about it.

Author: Added 1.
2 is part of Risks and Mitigations.

### Risks and Mitigations

* **Risk:** Having a routing preference like `PreferZone` comes at the risk of
  endpoints in certain zones being overloaded if the originating traffic is
  skewed towards a particular zone.

  **Mitigation:**
    * Emphasize in the documentation that the `PreferZone` preference is
      designed for low-latency or monetory-cost reasons, with the understanding
      that it can lead to overload within zones.
    * Recommend approaches like having deployments per zone which can scale
      independently of other zones.

* **Risk:** Users migrating from the `service.kubernetes.io/topology-mode`
  annotation might encounter differences in exact routing behavior:
  * The new field doesn't support a routing preference that is exactly similar
    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.

>   * If both field and the annotation are set, the annotation will take precedence.

Reviewer 2: ...for some number of releases, until it is deprecated and removed

Author: Added

  kube-controller-manager based implementations to say that we do support the
  standard heuristics.

### Changes within kube-proxy

**Present behaviour:** kube-proxy only considers EndpointSlice hints for route
programming if the `service.kubernetes.io/topology-aware-hints` annotation is
set to "Auto"

**New behaviour:** Irrespective of what the annotation
`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
set to (or even if they are not set at all), kube-proxy will always consider

> EndpointSlice hints.

Reviewer 2: "... kube-proxy will always consider EndpointSlice hints (assuming this feature gate is enabled)"

Author: Added!
